src/rust_connection/id_allocator.rs

```diff
@@ -42,24 +44,38 @@ impl IDAllocator {
         conn: &impl RequestConnection,
     ) -> Result<u32, ReplyOrIdError> {
         if self.next_id > self.max_id {
-            // TODO: Check if XC-MISC is available.
-
-            // Send an XC-MISC GetXIDRange request.
-            let xidrange = conn.xc_misc_get_xidrange()?.reply()?;
-            let (start, count) = (xidrange.start_id, xidrange.count);
-            // Apparently (0, 1) is how the server signals "I am out of IDs".
-            // The second case avoids an underflow below and should never happen.
-            if (start, count) == (0, 1) || count == 0 {
+            if self.check_xc_misc(conn)? {
+                // Send an XC-MISC GetXIDRange request.
+                let xidrange = conn.xc_misc_get_xidrange()?.reply()?;
+                let (start, count) = (xidrange.start_id, xidrange.count);
+                // Apparently (0, 1) is how the server signals "I am out of IDs".
+                // The second case avoids an underflow below and should never happen.
+                if (start, count) == (0, 1) || count == 0 {
+                    return Err(ReplyOrIdError::IdsExhausted);
+                }
+                self.next_id = start;
+                self.max_id = start + (count - 1) * self.increment;
+            } else {
                 return Err(ReplyOrIdError::IdsExhausted);
```

Comment:
Since I had a comment already anyway:

Could you reduce the indention a bit here? Add a `if !self.check_xc_misc(conn)? { return Err(ReplyOrIdError::IdsExhausted); }` at the top of this chunk and then the `else` can be removed:
```rust
            if self.check_xc_misc(conn)? {
                return Err(ReplyOrIdError::IdsExhausted);
            }
            // Send an XC-MISC GetXIDRange request.
            let xidrange = conn.xc_misc_get_xidrange()?.reply()?;
            let (start, count) = (xidrange.start_id, xidrange.count);
[...]
```

Reply:
Perhaps also rename `check_xc_misc` to `has_xc_misc`... and actually, why is that function necessary at all? The connection caches the `QueryExtensionReply` that it gives you from `extension_information`, so this whole cache doesn't really add much value, does it?

Reply:
And since I am writing lots of comments here anyway:

Can you replace the hardcoded `"XC-MISC"` with `crate::generated::xc_misc::X11_EXTENSION_NAME`?

Reply:
> Could you reduce the indention a bit here? Add a `if !self.check_xc_misc(conn)? { return Err(ReplyOrIdError::IdsExhausted); }` at the top of this chunk and then the `else` can be removed:

> Can you replace the hardcoded `"XC-MISC"` with `crate::generated::xc_misc::X11_EXTENSION_NAME`?



Will do it!

> Perhaps also rename `check_xc_misc` to `has_xc_misc`... and actually, why is that function necessary at all? The connection caches the `QueryExtensionReply` that it gives you from `extension_information`, so this whole cache doesn't really add much value, does it?

I created the function thinking that the extension would need some more initialization (e.g., something like a query version), but it doesn't seem to (at least it is not done in XCB). I will remove it and just use `extension_information`.